### ISSUE TITLE
Error in gradient of CTCLoss

### ISSUE THREAD

**Issue by danpovey** [pytorch/pytorch#127608]
### 🐛 Describe the bug

OK, I'm afraid I am describing from memory an issue that I more thoroughly investigated a few months ago and I wasn't able to dig up specific records or test-cases.  This is also something that my team ran into, and had to work around, a few years ago, and I don't believe it's changed.  The issue definitely exists in the CUDA version of ctc_loss, I don't recall whether it exists in the CPU version, I think it does not. [EDIT: I just checked, it does have the same issue, it refers to eqn. (16) in the paper, which is the problematic part.]

Basically, the CUDA version of torch's ctc_loss code acts as if it is renormalizing the log-probs.  That is: ctc_loss takes a logprobs arg, but it returns a gradient as if you had provided logprobs.log_softmax().  IIRC this is because the CUDA implementation of torch's ctc_loss was taken from a paper, and the paper described how to compute the CTC loss from un-normalized logprobs, but that is not how torch's interface works.  There is a certain equation in the paper, equation (16) in https://www.cs.toronto.edu/~graves/icml_2006.pdf , which is implemented in the CUDA code and doesn't have to be.  When I looked at the CUDA code it wasn't just a case of removing a couple lines though, it would require someone to study the code a bit, come up with a fix and test it.

This issue does not affect CTC training if the "logprobs" just comes from [something].log_softmax(), which it normally does, but there are various tricks that you might want to do, where this does affect the result and in fact gives a demonstrably wrong gradient that does not pass a gradient check: for example, if the logprobs are not properly normalized.

### Versions

I believe this affects all versions, I am just pasting the configuration currently on my mac so that it lets me submit the bug report.
```
OS: macOS 12.7.4 (x86_64)
GCC version: Could not collect
Clang version: 14.0.0 (clang-1400.0.29.202)
CMake version: version 3.15.4
Libc version: N/A

Python version: 3.9.0 (default, Nov 16 2020, 19:16:16)  [Clang 12.0.0 (clang-1200.0.32.27)] (64-bit runtime)
Python platform: macOS-12.7.4-x86_64-i386-64bit
Is CUDA available: N/A
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: Could not collect
Nvidia driver version: Could not collect
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: N/A

CPU:
Intel(R) Core(TM) i5-1038NG7 CPU @ 2.00GHz

Versions of relevant libraries:
[pip3] numpy==1.26.4
[pip3] onnx==1.10.1
[pip3] pytorch-lightning==1.3.8
[pip3] torch==1.8.1
[pip3] torch-scheduled-sampling==1.0.0
[pip3] torchaudio==0.8.1
[pip3] torchmetrics==0.4.1
[pip3] torchvision==0.9.1
[conda] Could not collect
```

cc @ezyang @albanD @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7 @mruberry @jbschlosser @walterddr @mikaylagawarecki

**Comment by mikaylagawarecki** [comment]
This seems related to https://github.com/pytorch/pytorch/issues/122243

The documentation for the  `log_softmax` argument does explicitly state that it should be obtained from the output of [`torch.nn.functional.log_softmax()`](https://pytorch.org/docs/stable/generated/torch.nn.functional.log_softmax.html#torch.nn.functional.log_softmax), so this behavior seems to be expected

We are happy to accept a PR that addresses this issue however

**Comment by soulitzer** [comment]
Closing duplicate in favor of https://github.com/pytorch/pytorch/issues/52241
